theforeman.org/pipelines/lib/packaging.groovy

```diff
@@ -63,3 +108,319 @@ def repoclosures(repo, releases, version) {
+    def package_dir
+
+    sh "mkdir -p ${build_dir}"
+
+    dir(build_dir) {
+        sh """
+            # Import variables from the project, allowing it to override behaviour
+            if [ -e ../${project}/build_vars.sh ]; then
+              . ../${project}/build_vars.sh
+            fi
+
+            if [[ \$BUILD_TYPE == "python" ]]; then
+              pip download --no-deps --no-binary=:all: "${project}==${package_version}"
+              mkdir "${project}-${package_version}"
+              tar -x -C "${project}-${package_version}" --strip-components=1 -f "${project}-${package_version}.tar.gz"
+              mv "${project}-${package_version}.tar.gz" "${project}_${package_version}.orig.tar.gz"
+              cp -r ../${project} "${project}-${package_version}/debian"
+            elif [[ \$BUILD_TYPE == "gem" ]]; then
+              gem fetch ${project} -v "=${package_version}"
+              ../../gem2deb ${project}-${package_version}.gem --debian-subdir ../${project} --only-source-dir
+            else
+              echo "Unsupported build type: \${BUILD_TYPE}"
+              exit 1
+            fi
+        """
+
+        package_dir = sh(returnStdout: true, script: "find -maxdepth 1 -type d -not -name '.'").trim()
+    }
+
+    dir("${build_dir}/${package_dir}") {
+        if (pull_request) {
+            last_commit = sh(reutrnStdout: true, script: "git rev-list HEAD|/usr/bin/head -n 1").trim()
+            add_debian_changelog(os, package_version, repoowner, last_commit)
+        }
+    }
+
+    return "${build_dir}/${package_dir}"
+}
+
+def setup_sources_plugin(project, os, version, repoowner, pull_request = false) {
+    def package_version
+    def build_dir = deb_build_dir('plugin', project)
+    def package_dir
+
+    sh "mkdir -p ${build_dir}"
+
+    dir(build_dir) {
+        if (project.contains('smart_proxy')) {
+            package_version = sh(returnStdout: true, script: "head -n1 ../${project}/changelog|awk '{print \$2}'|sed 's/(//;s/)//'|cut -f1 -d-").trim()
+
+            sh """
+                gem fetch ${project} -v "=${package_version}"
+                ../../dependencies/gem2deb ${project}-${package_version}.gem --debian-subdir ../${project} --only-source-dir
+            """
+
+            package_dir = sh(returnStdout: true, script: "find -maxdepth 1 -type d -not -name '.'").trim()
+        } else {
+            package_version = sh(returnStdout: true, script: "head -n1 ../${project}/debian/changelog|awk '{print \$2}'|sed 's/(//;s/)//'|cut -f1 -d-|cut -d: -f2").trim()
+
+            sh """
+                BUILD_TYPE=gem
+                # Import variables from the project, allowing it to override behaviour
+                if [ -e ../${project}/build_vars.sh ]; then
+                  . ../${project}/build_vars.sh
+                fi
+
+                if [ "\$BUILD_TYPE" = "gem" ]; then
+                    cp -r ../${project} ./
+                    cd ${project}
+                    ../../download_gems
```

Comment:
Anybody know why for plugins that are gems we run this download_gems script and for other gems (like dependencies) we use `gem fetch` and `gem2deb` ?

Reply:
I think (and I can be totally wrong) this is due to the totally different ways of how we build/vendor stuff.
* dependencies are "real" (mostly policy compliant) Debian packages with Ruby modules built against the system Ruby, so the workflow is: use `gem fetch` to obtain the `.gem` file, call `gem2deb` on it to convert it to a Debian (source) package, build it.
* core is a big bundle (generated by `bundle package`), with some system level dependencies from above.
* plugins are *extending* this (the core) bundle, to be able to do this, we pull down the gems (of the plugin and its dependencies, using `download_gems`) and place those in the package into the same folder the core bundle is installed (and a bundler.d snippet). on installation of the package, we call `bundle update` and thus actually add those modules to the bundle.

hope this helps, and didn't confuse you further?

Reply:
I think only Foreman plugins use bundler that way. Smart Proxy plugins are more policy compliant and AFAIK also use system Ruby (and system gems).

Reply:
That's of course correct and makes me no more happier ;)